dev_tools/lib/solidus/release_drafter/builder.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+require 'set'
+
+module Solidus
+  class ReleaseDrafter
+    # Defines the template for releases and adds entries to it.
+    #
+    # @api private
+    class Builder
+      TITLE_PREFIX = '## '
+
+      TITLE_TEMPLATE = lambda do |category|
+        TITLE_PREFIX + category
+      end.freeze
+
+      SEPARATOR = "\n"
+
+      TITLE_SEPARATOR = SEPARATOR
+
+      ENTRY_TEMPLATE = lambda do |title, number, user|
+        "- #{title} ##{number} (@#{user})"
+      end.freeze
+
+      ENTRY_SEPARATOR = SEPARATOR
+
+      SECTION_SEPARATOR = SEPARATOR + SEPARATOR
+
+      def initialize(draft:, categories:, prepend: '', append: '')
+        @categories = Set[*categories]
+        @prepend = prepend
+        @prepend_pattern = /^#{Regexp.quote(@prepend)}/m
+        @append = append
+        @append_pattern = /#{Regexp.quote(@append)}$/m
+        @draft = draft.new? ? draft.with(content: initial_content) : draft
+        sanitize_prepend_append
+        @ast = build_ast
+        sanitize_categories
+      end
+
+      def add(title:, number:, user:, categories:)
+        ENTRY_TEMPLATE.(title, number, user)
+          .then { |entry| add_to_ast(entry, categories) }
+          .then { |ast| @draft.with(content: unparse(ast)) }
+      end
+
+      private
+
+      def initial_content
+        @prepend + @categories.map do |category|
+          TITLE_TEMPLATE.(category)
+        end.join(SECTION_SEPARATOR) + @append
+      end
+
+      def build_ast
+        generated_lines.reduce([nil, {}]) do |(current_category, ast), line|
+          maybe_next_category = line.match(/^#{TITLE_PREFIX}(.*)$/)&.[](1)
+          if line.strip.empty?
+            [current_category, ast]
+          elsif maybe_next_category
+            [maybe_next_category, ast.merge(maybe_next_category => [])]
+          elsif current_category
+            [current_category, ast.merge(current_category => ast[current_category] + [line])]
+          else
+            [current_category, ast]
+          end
+        end[1]
+      end
```

Comment:
It seems the AbstractSyntaxTree methods (`build_ast`, `add_to_ast`, `unparse`) could be extracted to a separate AbstractSyntaxTree class. Doing so might simplify the Builder class and allow us to test the AbstractSyntaxTree methods directly. What do you think?

Reply:
I see pros & cons. For now, I'd prefer to leave it as it is as I'm worried about over-engineering it plus I already have spend a lot of time with this PR 😅